# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wgpu_bindings/src/client.rs
# Commit: dd10bf41fcea
# Full Hash: dd10bf41fcea6aa2f40d4528f4a7afc22b1f389e
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2024-04-05 21:43:02
# Regressor Bug: 1881518
# File Overlap Count: 1
# Description:
#   Bug 1881518 - Recycle indices again. r=webgpu-reviewers,bradwerth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D202433
# ==============================================================================

diff -r b2cba928a7bb -r dd10bf41fcea gfx/wgpu_bindings/src/client.rs
--- a/gfx/wgpu_bindings/src/client.rs	Fri Apr 05 07:28:07 2024 +0000
+++ b/gfx/wgpu_bindings/src/client.rs	Fri Apr 05 07:42:05 2024 +0000
@@ -525,6 +525,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_buffer_id(
+    client: &Client,
+    id: id::BufferId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .buffers
+        .free(id)
+}
+
+#[no_mangle]
 pub extern "C" fn wgpu_client_create_texture(
     client: &Client,
     device_id: id::DeviceId,
@@ -555,6 +569,21 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_texture_id(
+    client: &Client,
+    id: id::TextureId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .textures
+        .free(id)
+}
+
+
+#[no_mangle]
 pub extern "C" fn wgpu_client_create_texture_view(
     client: &Client,
     device_id: id::DeviceId,
@@ -590,6 +619,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_texture_view_id(
+    client: &Client,
+    id: id::TextureViewId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .texture_views
+        .free(id)
+}
+
+#[no_mangle]
 pub extern "C" fn wgpu_client_create_sampler(
     client: &Client,
     device_id: id::DeviceId,
@@ -624,6 +667,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_sampler_id(
+    client: &Client,
+    id: id::SamplerId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .samplers
+        .free(id)
+}
+
+#[no_mangle]
 pub extern "C" fn wgpu_client_make_encoder_id(
     client: &Client,
     device_id: id::DeviceId,
@@ -639,6 +696,21 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_command_encoder_id(
+    client: &Client,
+    id: id::CommandEncoderId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .command_buffers
+        .free(id.transmute())
+}
+
+
+#[no_mangle]
 pub extern "C" fn wgpu_client_create_command_encoder(
     client: &Client,
     device_id: id::DeviceId,
@@ -700,6 +772,7 @@
     }
 }
 
+
 #[no_mangle]
 pub unsafe extern "C" fn wgpu_render_bundle_encoder_destroy(
     pass: *mut wgc::command::RenderBundleEncoder,
@@ -755,6 +828,20 @@
     id
 }
 
+#[no_mangle]
+pub extern "C" fn wgpu_client_free_render_bundle_id(
+    client: &Client,
+    id: id::RenderBundleId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .render_bundles
+        .free(id)
+}
+
 #[repr(C)]
 pub struct ComputePassDescriptor<'a> {
     pub label: Option<&'a nsACString>,
@@ -986,6 +1073,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_bind_group_layout_id(
+    client: &Client,
+    id: id::BindGroupLayoutId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .bind_group_layouts
+        .free(id)
+}
+
+#[no_mangle]
 pub unsafe extern "C" fn wgpu_client_render_pipeline_get_bind_group_layout(
     client: &Client,
     pipeline_id: id::RenderPipelineId,
@@ -1059,6 +1160,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_pipeline_layout_id(
+    client: &Client,
+    id: id::PipelineLayoutId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .pipeline_layouts
+        .free(id)
+}
+
+#[no_mangle]
 pub unsafe extern "C" fn wgpu_client_create_bind_group(
     client: &Client,
     device_id: id::DeviceId,
@@ -1106,6 +1221,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_bind_group_id(
+    client: &Client,
+    id: id::BindGroupId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .bind_groups
+        .free(id)
+}
+
+#[no_mangle]
 pub extern "C" fn wgpu_client_make_shader_module_id(
     client: &Client,
     device_id: id::DeviceId,
@@ -1120,6 +1249,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_shader_module_id(
+    client: &Client,
+    id: id::ShaderModuleId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .shader_modules
+        .free(id)
+}
+
+#[no_mangle]
 pub unsafe extern "C" fn wgpu_client_create_compute_pipeline(
     client: &Client,
     device_id: id::DeviceId,
@@ -1161,6 +1304,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_compute_pipeline_id(
+    client: &Client,
+    id: id::ComputePipelineId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .compute_pipelines
+        .free(id)
+}
+
+#[no_mangle]
 pub unsafe extern "C" fn wgpu_client_create_render_pipeline(
     client: &Client,
     device_id: id::DeviceId,
@@ -1204,6 +1361,20 @@
 }
 
 #[no_mangle]
+pub extern "C" fn wgpu_client_free_render_pipeline_id(
+    client: &Client,
+    id: id::RenderPipelineId,
+) {
+    let backend = id.backend();
+    client
+        .identities
+        .lock()
+        .select(backend)
+        .render_pipelines
+        .free(id)
+}
+
+#[no_mangle]
 pub unsafe extern "C" fn wgpu_command_encoder_copy_buffer_to_buffer(
     src: id::BufferId,
     src_offset: wgt::BufferAddress,
